### ISSUE TITLE
Annyang stops working after a bit of silence

### ISSUE THREAD

**Issue by sanjayshreedharan** [TalAter/annyang#206]
Annyang works perfectly fine, when it is started for the first time. But, after some time (silence), it doesn't seem to recognize what the user is speaking. Leave it for some more time and it again starts recognising but it recognises what was previously spoken as well. 

Also, when I add callback for error, I was able to see error popping up every now and then. Is there a way to possibly know what error it is ?

Environment :
* Browser Name and version: Chrome (version 48)
* Operating System and version (desktop or mobile): Ubuntu Mate 15.04

**Comment by lmangani** [comment]
@sanjayshreedharan hard to tell if you don't share the location of your callback or what error was received. Could you extend with more information? Also what's the system load at the time this happens? You might be experiencing a local queuing/load side-effect of some kind since you still receive the previous keywords apparently, suggesting Annyang not being the problem.

**Comment by sanjayshreedharan** [comment]
Just added these lines to my script.

```
annyang.addCallback('error', function() {
                console.log('There was an error in Annyang!');
   });

```

After adding these lines, an error pops up in the console every now and then. 

This problem occurs even when there is only one Chrome tab which is open and no other applications are running. So, am pretty sure it is not a problem of loading.

**Comment by lmangani** [comment]
how about
```
annyang.addCallback('error', function(err) {
                console.log('There was an error in Annyang!',err);
   });
```

**Comment by sanjayshreedharan** [comment]
@lmangani The error is printed as undefined. There is no variable that is being returned, and hence it shows as undefined.

**Comment by lmangani** [comment]
that's unfortunate - are you able to reproduce this issue with different client device/OS combinations?

**Comment by sanjayshreedharan** [comment]
I've tried this with OSX El Captain 10.11.5 and was able to reproduce the issue.

**Comment by sanjayshreedharan** [comment]
When I tried logging the events in recognition.onerrror, I received an error. I have attached the screenshot for reference. 

![image](https://cloud.githubusercontent.com/assets/13719639/15706688/7d72e684-2812-11e6-968a-f2456ed0d7c0.png)

**Comment by stilltrackin** [comment]
I'm seeing this same behavior on OSX El Capitan 10.11.5 with Chrome 50.0.2661.102 (64-bit).  It works fine while I'm using it, but if there is a minute of silence it stops responding.

**Comment by TalAter** [comment]
Could you upload some code so we can see it? And please provide as many details as possible. For example, are you hosting the sample on http, https, localhost, file?

**Comment by sanjayshreedharan** [comment]
<script src='http://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.3/jquery.min.js'></script>
    <script src="http://cdnjs.cloudflare.com/ajax/libs/annyang/2.4.0/annyang.min.js"></script>
    <script>
        if (annyang)
        {
            annyang.debug();
            annyang.start();
            setInterval( function()
            {
                listen = annyang.isListening()
                console.log(listen);   
            } , 1000);
            annyang.addCallback('error', function(err) {
                //$('.myErrorText').text('There was an error!');
                console.log('There was an error in Annyang!',err);
        });
        }
        else
            console.log("Annyang not working")

This is hosted on localhost.

**Comment by anhlee24** [comment]
i have the same issue, annyang was not listen after a bit of slience (about 1m or longer). No error show.
When i check listening by annyang.isListening(), it return TRUE.
Have you found the reason?

Thank you.

**Comment by alecbenson** [comment]
Getting the same problem... Anyone know what's up?

**Comment by mwkrieger** [comment]
Hi, I'm seeing this happen as well. The error that is firing is "no-speech". 

Has anyone found a workaround?

**Comment by JVRibeiro** [comment]
Maybe attach annyang.start() method to the error callback is the solution. Something like this:

annyang.addCallback('error', function ()
{
if ( !annyang.isListening() ) {
annyang.start();
}
});

**Comment by TalAter** [comment]
@JVRibeiro That is a very dangerous approach.

The error might be from something else like a user declining the permission, the device not having a microphone (e.g. desktop pc), etc. In such a case, restarting the speech recognition might be followed by an immediate abort and an error, which would in turn start it again, followed by another error and so on... This could cause the browser to go into an endless loop, which might cause the tab, or even the entire browser, to crash. I've detailed this in this Chromium bug report - https://bugs.chromium.org/p/chromium/issues/detail?id=296690

In fact, annyang already does something like what you are suggesting, but in a way that makes sure such an infinite loop won't happen. I've added details on this below.

**Comment by TalAter** [comment]
**tl;dr**
The `no-speech` error is triggered by the browser automatically. You can work around it with annyang's `autorestart` mode.

----

When the browser doesn't detect any speech for a while, it triggers a `no-speech` error, and turns off the speech recognition.

When annyang is in `autorestart` mode, it will automatically restart the speech recognition engine if it is stopped by this event.

`autorestart` is on by default when using HTTPS, and off by default on HTTP.

You can force annyang to start in either mode by passing it in the options object:
````javascript
// Start listening, restart automatically
annyang.start({ autoRestart: true });
````

This is detailed in the docs for [`annyang.start()`](https://github.com/TalAter/annyang/tree/master/docs#startoptions)
